### PR TITLE
Fixed Go hook registration

### PR DESCRIPTION
**Pull request by zuzzas**
#### What this PR does / why we need it

- fix panic caused by the absent rate limiter
- Metadata is calculated on sdk.Register call
- simplification: sdk.Register call and a hook's structure should be in the same file

#### Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: sdk/registry.go [hunk 1]
```diff
@@ -1,11 +1,26 @@
...
+	"runtime"
 	"sync"
 
 	"github.com/flant/addon-operator/pkg/module_manager/go_hook"
 )
 
+// /path/.../global-hooks/a/b/c/Hook-name.go
+// $1 - Hook path
+// $3 - Hook name
+var globalRe = regexp.MustCompile(`/global-hooks/(([^/]+/)*([^/]+))$`)
```
Comment (diafour) [inline sdk/registry.go:14]:
```suggestion
var globalRe = regexp.MustCompile(`/global[\/\-]hooks/(([^/]+/)*([^/]+))$`)
```

There are 2 layouts:

1. All hooks in `/global-hooks`
2. Hooks in `/global/hooks` and OpenAPI schemas in `/global/openapi` — there was a plan to use it.

File: sdk/registry.go [hunk 2]
```diff
@@ -37,12 +57,63 @@ func Registry() *HookRegistry {
...
+	hookMeta := &go_hook.HookMetadata{
+		Name:       "",
+		Path:       "",
+		Global:     false,
+		Module:     false,
+		ModuleName: "",
+	}
+
+	pc := make([]uintptr, 50)
+	n := runtime.Callers(0, pc)
```
[the remark this answers is not in the document]
Comment (zuzzas) [reply sdk/registry.go:77]: Will you vouch that the call stack (an ABI in a sense) will not change over time?

File: sdk/registry.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (zuzzas) [reply sdk/registry.go]: Won't work.
https://github.com/flant/addon-operator/pull/187#discussion_r606655380

File: sdk/registry.go [hunk 3]
```diff
@@ -37,12 +57,63 @@ func Registry() *HookRegistry {
...
+func (h *HookRegistry) Hooks() []HookWithMetadata {
 	return h.hooks
 }
 
 func (h *HookRegistry) Add(hook go_hook.GoHook) {
 	h.m.Lock()
 	defer h.m.Unlock()
-	h.hooks = append(h.hooks, hook)
+
+	hookMeta := &go_hook.HookMetadata{
```
[the remark this answers is not in the document]
Comment (zuzzas) [reply sdk/registry.go]: Of course. It was copy-pasted. Fixed!

File: sdk/registry.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (nabokihms) [reply sdk/registry.go:71]: I don't know. I was just wondering.

File: sdk/registry.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (nabokihms) [reply sdk/registry.go]: Frankly, I did not get the idea of the comment.
Do we have two hardcoded directories for the global hooks? It seems to be a problem.

Could we open an issue about it? Global and modules dirs should be configurable. We will solve this on v2 refactoring.

File: sdk/registry.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (zuzzas) [reply sdk/registry.go:71]: That's precisely my point. If we were to change a call stack by introducing a function somewhere, we'll bork the registration process.

File: sdk/registry.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (zuzzas) [reply sdk/registry.go]: https://github.com/flant/addon-operator/issues/188

Can you create a v2 milestone and start adding issues (and, subsequently, PRs) to it?
@diafour /cc

[the remark this answers is not in the document]
Comment (nabokihms) [reply sdk/registry.go]: https://github.com/flant/addon-operator/milestone/9

File: sdk/registry.go [hunk 4]
```diff
@@ -1,11 +1,26 @@
...
+// $3 - Hook name
+var globalRe = regexp.MustCompile(`/global[\/\-]hooks/(([^/]+/)*([^/]+))$`)
+
+// /path/.../modules/module-name/hooks/a/b/c/Hook-name.go
+// $1 - Hook path
+// $2 - module name
+// $4 - Hook name
+var moduleRe = regexp.MustCompile(`/modules/(([^/]+)/hooks/([^/]+/)*([^/]+))$`)
+
+var moduleNameRe = regexp.MustCompile(`^[0-9][0-9][0-9]-(.*)$`)
```
Comment (nabokihms) [inline sdk/registry.go:22]: I assume that this regex is incorrect because we have no rule for a module to start with a number.
Could you add a TODO here?

**Review by nabokihms — APPROVED** [review]

File: sdk/registry.go (hunk 4 above)
Comment (zuzzas) [reply sdk/registry.go:23]: Done.

Comment (diafour) [reply sdk/registry.go:23]: There is an issue about these "ugly hardcoded" numbers: https://github.com/flant/addon-operator/issues/39 (also #7)

**Review by diafour — APPROVED** [review]
lgtm

**Pull request merged**